src/appleseed/foundation/math/knn/knn_anyquery.h

```diff
@@ -0,0 +1,397 @@
+        const VectorType* APPLESEED_RESTRICT point_end = point_ptr + node->get_point_count();
+
+        while (point_ptr < point_end)
+        {
+            FOUNDATION_KNN_QUERY_STATS(++tested_point_count);
+
+            const ValueType square_dist = square_distance(*point_ptr++, query_point);
+
+            if (square_dist <= query_max_square_distance)
+                goto found;
+        }
+    }
+
+    //
+    // Step 3:
+    //
+    //   Traverse again the tree, starting at the top, but this time we will push
+    //   nodes that we don't visit (but that potentially contain candidate points)
+    //   to a partially-ordered stack.
+    //
+
+#define ORDER_NODE_ENTRIES(LhsIndex, RhsIndex)                                                  \
+    if (node_queue[node_queue_size - (LhsIndex)].m_dvec_square_norm <                           \
+        node_queue[node_queue_size - (RhsIndex)].m_dvec_square_norm)                            \
+    {                                                                                           \
+        const NodeEntry tmp = node_queue[node_queue_size - (LhsIndex)];                         \
+        node_queue[node_queue_size - (LhsIndex)] = node_queue[node_queue_size - (RhsIndex)];    \
+        node_queue[node_queue_size - (RhsIndex)] = tmp;                                         \
+    }
+
+    const size_t NodeQueueSize = 128;
+    NodeEntry node_queue[NodeQueueSize];
+    size_t node_queue_size = 0;
+
+    // Start at the root.
+    node = nodes;
+
+    FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+    while (node->is_interior())
+    {
+        const size_t split_dim = node->get_split_dim();
+        const ValueType split_abs = node->get_split_abs();
+        const ValueType split_dist = query_point[split_dim] - split_abs;
+
+        // Figure out which child node to follow (the one that contains the query point) and which one to push.
+        // Points on the split plane belong to the right child node.
+        const size_t follow_index = split_dist >= ValueType(0.0) ? 1 : 0;
+        const NodeType* APPLESEED_RESTRICT left_child_node = nodes + node->get_child_node_index();
+        const NodeType* APPLESEED_RESTRICT follow_node = left_child_node + follow_index;
+        const NodeType* APPLESEED_RESTRICT queue_node = left_child_node + 1 - follow_index;
+
+        FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+        // Like in the initial step, we stop as soon as we reached a node with enough points.
+        if (follow_node->get_point_count() < IdealLeafSize)
+            break;
+
+        // Push the node that we don't visit now to the node stack.
+        const ValueType square_split_dist = square(split_dist);
+        if (square_split_dist < query_max_square_distance)
+        {
+            VectorType dvec(0.0);
+            dvec[split_dim] = split_dist;
+
+            // Push the node to the node stack.
```

Comment:
The node array, which is called `node_queue` gets referred here to as a node stack. I'd recommend unified naming.

Reply:
Well ok, you're right, it's actually a partially ordered stack :) Will stick to stack then.